### PR TITLE
Support running Netty (in particular netty-tcnative) in the bootstrap class loader

### PR DESCRIPTION
**Pull request by trask**
Motivation:

Fix NullPointerExceptions that occur when running netty-tcnative inside the bootstrap class loader.

Modifications:

Replace `loader.getResource(...)` with `ClassLoader.getSystemResource(...)` when loader is null.

Replace `loader.loadClass(...)` with `Class.forName(..., false, loader)` which works when loader is both null and non-null.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trask I wonder if we could somehow have an unit test.

**Comment by ejona86** [comment]
For the test, you could add a binary to the bootstrap class loader, but just as a weird name "test-bs-loading.so" and load it in the test. But that'd be pretty annoying.

**Comment by normanmaurer** [comment]
@ejona86 @trask if its too hard to add a test than dont worry... I just would prefer to have a unit test so we ensure we not break it again in the future.

@trask let me know what you think.

**Comment by trask** [comment]
@normanmaurer @ejona86 unit test added, let me know what you think, the test caught one additional NPE, see additional change in latest push:

```
if (loader == null) {
    // cannot defineClass inside bootstrap class loader
    throw e1;
}
```

**Comment by ejona86** [comment]
Oh, hey! _Much_ easier test. Right! The file doesn't even need to exist.

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (58e74e9fee1c12eb3a73c38109cfd081750992b6) and 4.0 (b12a1e3fa66b61a7ed625b2d4ee2c901171a2074).

@trask thanks!

**Comment by trask** [comment]
Thanks for the quick review and merge

File: common/src/test/java/io/netty/util/internal/NativeLibraryLoaderTest.java [hunk 1]
```diff
@@ -25,13 +26,56 @@
 
 public class NativeLibraryLoaderTest {
 
+    private static final Method getSupressedMethod = getGetSuppressed();
```
Comment (carl-mastrangelo) [inline common/src/test/java/io/netty/util/internal/NativeLibraryLoaderTest.java]: ThrowableUtil might have been a better place for this.

Comment (trask) [reply common/src/test/java/io/netty/util/internal/NativeLibraryLoaderTest.java]: I thought it was better not to put test-only code (getSuppressed) inside of non- test code (ThrowableUtil), but if that's not a concern I can move it there.

**Pull request closed without merging**